Problem: Read/Write Transaction algorithm is incorrect

omarkj opened the pull request:
Solution: Use a queue to make transaction picker straight forward and easier to reason about.

This uncovered some badly defined doc tests where tests that now require a transaction were not in one. That has been taken care of.

https://github.com/PumpkinDB/PumpkinDB/issues/160

Commit:
Problem: Transaction ordering is undetermined

Solution: Introduce a queue that contains the transaction,
and ensure that transactins are operated on in the correct
order.

This removes the need for some checking code, as those
checks can now be done while popping transactions of the
queue.

Commit:
Problem: Doctests do not enforce transaction ordering

Solution: Add missing transactions to doctest.

yrashk commented on src/script/queue.rs:
```diff
@@ -0,0 +1,81 @@
+#[derive(Debug)]
```
Any reason for not using [VecDeque](https://doc.rust-lang.org/1.15.0/std/collections/struct.VecDeque.html)? It doesn't have `peek` but it can be easily simulated with `get`

omarkj replied:
Decided against it since it was badly documented what happens when you push to it when it's full. It's documented as a ring buffer, so I figured it would override the oldest elements in the queue which I do not want.

yrashk replied:
It will just reallocate to push more, its growable: "VecDeque is a growable ring buffer"

yrashk commented on src/script/mod_storage.rs:
```diff
@@ -20,14 +20,18 @@ use std::collections::HashMap;
 use byteorder::{BigEndian, WriteBytesExt};
 use snowflake::ProcessUniqueId;
+use std::collections::BTreeMap;
+use script::queue::{Queue};
```
Nitpick: rustfmt will convert this into `::Queue` on the next run.

yrashk commented on src/script/mod_storage.rs:
```diff
@@ -20,14 +20,18 @@ use std::collections::HashMap;
 pub type CursorId = ProcessUniqueId;
 
+const STACK_EMPTY_CLOSURE: &'static [u8] = b"";
+const MAX_TXNS_PER_ENV: usize = 64;
```
Should we document this limit somewhere (WRITE?)

yrashk commented on src/script/mod_storage.rs:
```diff
@@ -20,14 +20,18 @@ use std::collections::HashMap;
+
 instruction!(WRITE, b"\x85WRITE");
-instruction!(WRITE_END, b"\x80\x85WRITE"); // internal instruction
+instruction!(WRITE_END, b"\x80\x85WRITE"); // internal word
```
should be `// internal instruction`, not `// internal word`

Commit:
Problem: Number of transaction is not documented

Solution: Document max number of transactions
in the files that pertain to transaction opening.

yrashk commented:
Also, I think it would be valuable to bring the test case from #181 (possibly needs an improvement):

```
PumpkinDB> "hello" HLC CONCAT DUP ["world" ASSOC [RETR] READ] WRITE.
"world"
```

Basically, we need to write a case in which we can validate that a proper transaction is picked. In the above transaction, `RETR` actually goes into the `WRITE` txn scope, while it should have went into a `READ`.

This way we can ensure the order.

Commit:
Problem: No test for new txn order guarantees

Solution: Add a test to test for new txn order
behaviour

yrashk commented:
Awesome! ❤️
